// gl-react/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Bus", {
  enumerable: true,
  get: function () {
    return _Bus.default;
  }
});
Object.defineProperty(exports, "connectSize", {
  enumerable: true,
  get: function () {
    return _connectSize.default;
  }
});
Object.defineProperty(exports, "createSurface", {
  enumerable: true,
  get: function () {
    return _createSurface.default;
  }
});
Object.defineProperty(exports, "listSurfaces", {
  enumerable: true,
  get: function () {
    return _createSurface.list;
  }
});
Object.defineProperty(exports, "GLSL", {
  enumerable: true,
  get: function () {
    return _GLSL.default;
  }
});
Object.defineProperty(exports, "LinearCopy", {
  enumerable: true,
  get: function () {
    return _LinearCopy.default;
  }
});
Object.defineProperty(exports, "NearestCopy", {
  enumerable: true,
  get: function () {
    return _NearestCopy.default;
  }
});
Object.defineProperty(exports, "Node", {
  enumerable: true,
  get: function () {
    return _Node.default;
  }
});
Object.defineProperty(exports, "Shaders", {
  enumerable: true,
  get: function () {
    return _Shaders.default;
  }
});
Object.defineProperty(exports, "Uniform", {
  enumerable: true,
  get: function () {
    return _Uniform.default;
  }
});
Object.defineProperty(exports, "Visitor", {
  enumerable: true,
  get: function () {
    return _Visitor.default;
  }
});
Object.defineProperty(exports, "VisitorLogger", {
  enumerable: true,
  get: function () {
    return _VisitorLogger.default;
  }
});
Object.defineProperty(exports, "Visitors", {
  enumerable: true,
  get: function () {
    return _Visitors.default;
  }
});
exports.Backbuffer = void 0;

var _Bus = _interopRequireDefault(require("./Bus"));

var _connectSize = _interopRequireDefault(require("./connectSize"));

var _createSurface = _interopRequireWildcard(require("./createSurface"));

var _GLSL = _interopRequireDefault(require("./GLSL"));

var _LinearCopy = _interopRequireDefault(require("./LinearCopy"));

var _NearestCopy = _interopRequireDefault(require("./NearestCopy"));

var _Node = _interopRequireDefault(require("./Node"));

var _Shaders = _interopRequireDefault(require("./Shaders"));

var _Uniform = _interopRequireDefault(require("./Uniform"));

var _Visitor = _interopRequireDefault(require("./Visitor"));

var _VisitorLogger = _interopRequireDefault(require("./VisitorLogger"));

var _Visitors = _interopRequireDefault(require("./Visitors"));

require("webgltexture-loader-ndarray");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// DEPRECATED
const Backbuffer = "Backbuffer";
exports.Backbuffer = Backbuffer;
