// webgltexture-loader/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "WebGLTextureLoader", {
  enumerable: true,
  get: function () {
    return _WebGLTextureLoader.default;
  }
});
Object.defineProperty(exports, "WebGLTextureLoaderAsyncHashCache", {
  enumerable: true,
  get: function () {
    return _WebGLTextureLoaderAsyncHashCache.default;
  }
});
Object.defineProperty(exports, "WebGLTextureLoaderSyncHashCache", {
  enumerable: true,
  get: function () {
    return _WebGLTextureLoaderSyncHashCache.default;
  }
});
Object.defineProperty(exports, "LoadersRegistry", {
  enumerable: true,
  get: function () {
    return _LoadersRegistry.default;
  }
});
Object.defineProperty(exports, "LoaderResolver", {
  enumerable: true,
  get: function () {
    return _LoaderResolver.default;
  }
});
Object.defineProperty(exports, "globalRegistry", {
  enumerable: true,
  get: function () {
    return _globalRegistry.default;
  }
});

var _WebGLTextureLoader = _interopRequireDefault(require("./WebGLTextureLoader"));

var _WebGLTextureLoaderAsyncHashCache = _interopRequireDefault(require("./WebGLTextureLoaderAsyncHashCache"));

var _WebGLTextureLoaderSyncHashCache = _interopRequireDefault(require("./WebGLTextureLoaderSyncHashCache"));

var _LoadersRegistry = _interopRequireDefault(require("./LoadersRegistry"));

var _LoaderResolver = _interopRequireDefault(require("./LoaderResolver"));

var _globalRegistry = _interopRequireDefault(require("./globalRegistry"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
